### ISSUE TITLE
Add Deis scheduler

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#280]
**Is your feature request related to a problem? Please describe.**

State of the art scheduler is missing.

**Describe the solution you'd like**

DEIS scheduler should be integrated: https://github.com/qsh-zh/deis

**Comment by rashmimarganiatgithub** [comment]
@patrickvonplaten  I can take up this issue , can you please assign it to me?

**Comment by patrickvonplaten** [comment]
Sure! Count my message here as assigning the issue to you ;-) Let me know if you need any help! 
A couple of pointers from my side as I think this might be a tougher issue/PR:

I would suggest the following approach to add DEIS to `diffusers`:
- 1. Get this code snippet working in PyTorch: https://github.com/qsh-zh/deis#if-diffusion-models-are-trained-with-continuous-time where as you can use a model trained on continous timesteps from the Hub: https://huggingface.co/google/ncsnpp-celebahq-256

More specifically in a first step we should try to get this to work:

```python
from diffusers import UNet2DModel

model = UNet2DModel.from_pretrained("google/ncsnpp-celebahq-256")

# now the code of https://github.com/qsh-zh/deis#if-diffusion-models-are-trained-with-continuous-time

import torch as th
import th_deis as deis
def eps_fn(x_t, scalar_t):
    vec_t = (th.ones(x_t.shape[0])).float().to(x_t) * scalar_t
    with th.no_grad():
        return eps_model(x_t, vec_t)

# mappings between t and alpha in VPSDE
# we provide popular linear and cos mappings
t2alpha_fn,alpha2t_fn = deis.get_linear_alpha_fns(beta_0=0.01, beta_1=20)

vpsde = deis.VPSDE(
    t2alpha_fn, 
    alpha2t_fn,
    sampling_eps, # sampling end time t_0
    sampling_T # sampling starting time t_T
)

sampler_fn = deis.get_sampler(
    # args for diffusion model
    vpsde,
    eps_fn,
    # args for timestamps scheduling
    ts_phase="t", # support "rho", "t", "log"
    ts_order=2.0,
    num_step=10,
    # deis choice
    method = "t_ab", # deis sampling algorithms: support "rho_rk", "rho_ab", "t_ab", "ipndm"
    ab_order= 3, # greater than 0, used for "rho_ab", "t_ab" algorithms, other algorithms will ignore the arg
    rk_method="3kutta" # used for "rho_rk" algorithms, other algorithms will ignore the arg
)

sample = sampler_fn(noise)
```

I haven't tested this code, but in a first step we should try to get this running with the original code base.
Maybe the authors (cc @qsh-zh and @jiamings ) could potentially help here as well :-)

- 2. In the next step, we should then move all the sampler logic to `diffusers` and get it working with diffusers code. 
- 3. Finally we should create a nice pipeline and link to the paper

Let me know if you have any questions @rashmimarganiatgithub

**Comment by jiamings** [comment]
Hi! Glad to help, although I am only an author on the DDIM paper, not the DEIS one.

I think a big problem with the DEIS codebase is that it depends on jax to compute the coefficients-- installing jax cpu would work as we only need the weights, but I am not sure if is desired for HuggingFace repos. 

I tried to implement this with pytorch, but it is very hard as functorch is not nearly as well-developed as jax -- I had to write my own vmap function, and even then the coefficients are slightly (1e-4 level) off from the ones from DEIS.

But if jax is not a problem, then the algorithm is actually quite straightforward after you get the coefficients.

**Comment by qsh-zh** [comment]
Hi! Gad to help.

There are three 4 new algorithms in DEIS paper, `tAB-DEIS`, `$\rho$AB-DEIS`, `$\rho$RK-DEIS`, and improved PNDM.

1. `$\rho$ AB-DEIS` and `$\rho$RK-DEIS` should be straightforward to integrate as it only requires two changes of variables then using standard RK and multistep solvers. Here is the impl for two transformations in VPSDE [t2rho](https://github.com/qsh-zh/deis/blob/6f31cf2d271119312b10bc10d60d76369eb024bd/th_deis/vpsde.py#L66) and [x2y](https://github.com/qsh-zh/deis/blob/main/th_deis/vpsde.py#L76).
2. Compared with PNDM, improved PNDM use lower order multistep for warming up for the first very iterations. I actually find diffuser repo has integrated the method in [ldm](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lms_discrete.py#L116).
3. As @jiamings  mentioned, `tAB-DEIS` requires a numerical integration [eq-15](https://arxiv.org/pdf/2204.13902.pdf). DEIS repo replies on `jax vmap` to calculate those coefficients.  If you want to get rid of jax, `scipy.integrate` is another option for numerical integration.

**Comment by patrickvonplaten** [comment]
@rashmimarganiatgithub note that this is quite a difficult contribution - so feel free to ask for help from the community!

**Comment by rashmimarganiatgithub** [comment]
@patrickvonplaten , will reach to community once I start working on this issue, but before let me close other easily issues in all possible way. I am working on it.

**Comment by rashmimarganiatgithub** [comment]
Working on it

**Comment by rashmimarganiatgithub** [comment]
Hi @qsh-zh, 

After going through your [code](https://github.com/qsh-zh/deis/blob/main/tests/test_tab_deis.py), I was able to modify the above-mentioned eps_fn()  without eps_model() with some dummy data as follows, 

data_shape = (3, 13)
noise = th.ones(data_shape)

def th_eps_fn(x, t):
    del t
    return th.ones_like(x)

# mappings between t and alpha in VPSDE
# we provide popular linear and cos mappings
t2alpha_fn, alpha2t_fn = th_deis.get_linear_alpha_fns(0.01, 20)

exp_sde = th_deis.VPSDE(t2alpha_fn, alpha2t_fn, 1e-3, 1.0)
sampler_fn = th_deis.get_sampler(
        exp_sde,
        th_eps_fn,
        ts_phase="t",
        ts_order=2.0,
        num_step=10,
        method = "ipndm",
    )
    
sample = sampler_fn(noise)

I went through all the files within the [th_deis folder](https://github.com/qsh-zh/deis/blob/main/th_deis) and none of the files contain eps_model() code. Can you please throw some light so that I can have the reference for eps_model() implementation and can you please provide me with some demo examples using this sampler_fn with some model training? which can be very helpful for me to move a bit quicker.

Thanks.

**Comment by qsh-zh** [comment]
@rashmimarganiatgithub  here is [one demo](https://github.com/qsh-zh/deis/blob/main/demo/discrete_celeba/runner/runner.py#L185-L214) in celeba based on PNDM codebase.

`eps_fn` is for predicting noise added to noised data. For example, the loss function of continuous DDPM follows, where the first argument of epsilon model is noise data `x_t`
![image](https://user-images.githubusercontent.com/28699575/189497524-6bee2dc9-51c7-4a85-98bb-3afbad1ae0e2.png)

**Comment by rashmimarganiatgithub** [comment]
@qsh-zh, thanks for pointing out on PNDM codebase and for clarifying my doubt.

**Comment by rashmimarganiatgithub** [comment]
@@qsh-zh, is this [sampler.py](https://github.com/qsh-zh/deis/blob/main/th_deis/sampler.py) only file which contains all sampler logic for th_deis, is there any dev doc/paper to understand about all the sampler functions?

**Comment by qsh-zh** [comment]
@rashmimarganiatgithub  you can check the paper, https://arxiv.org/abs/2204.13902

**Comment by williamberman** [comment]
@patrickvonplaten can we close issue since https://github.com/huggingface/diffusers/pull/1432 was merged?
